rts/lra/lra-test/lra-test-arquillian-extension/src/main/java/io/narayana/lra/arquillian/spi/LRAAnnotationAdjuster.java

```diff
@@ -0,0 +1,206 @@
+                Method method = Class.class.getDeclaredMethod(ANNOTATION_DATA_METHOD_NAME, null);
+                method.setAccessible(true);
+                // AnnotationData is private need to work with Object
+                Object annotationData = method.invoke(clazzToLookFor);
+                // AnnotationData works with map annotations
+                Field annotations = annotationData.getClass().getDeclaredField(ANNOTATIONS_FIELD_NAME);
+                annotations.setAccessible(true);
+                Map<Class<? extends Annotation>, Annotation> map =
+                        (Map<Class<? extends Annotation>, Annotation>) annotations.get(annotationData);
+                log.debugf("Adjusting LRA annotation %s for class %s%n", originalLRAAnnotation, clazzToLookFor.getName());
+                map.put(LRA.class, new LRAWrapped(originalLRAAnnotation));
+            } catch (Exception  e) {
+                throw new IllegalStateException("Cannot change annotation " + originalLRAAnnotation
+                        + " of class " + clazzToLookFor, e);
+            }
+        }
+    }
+
+    /**
+     * Changing the LRA annotation declared on method by wrapping it with {@link LRAWrapped}.
+     */
+    static void adjustLRAAnnotation(Method method, LRA originalLRAAnnotation) {
+        Field field = null;
+        try {
+            Class<?> executableClass = Class.forName("java.lang.reflect.Executable");
+            field = executableClass.getDeclaredField("declaredAnnotations");
+            field.setAccessible(true);
+        } catch (ClassNotFoundException cnfe) {
+            throw new IllegalStateException("Cannot instantiate class java.lang.reflect.Executable", cnfe);
+        } catch (NoSuchFieldException nsfe) {
+            throw new IllegalStateException("Cannot find field 'declaredAnnotations' under instantiate class java.lang.reflect.Executable", nsfe);
+        }
+        Map<Class<? extends Annotation>, Annotation> annotations = null;
+        try {
+            annotations = (Map<Class<? extends Annotation>, Annotation>) field.get(method);
+        } catch (IllegalAccessException iae) {
+            throw new IllegalStateException("Cannnot access field 'declaredAnnotations' of the method instance " + method, iae);
+        }
+        log.debugf("Adjusting LRA annotation %s for method %s of class %s%n",
+                originalLRAAnnotation, method, method.getDeclaringClass().getName());
+        annotations.put(LRA.class, new LRAWrapped(originalLRAAnnotation));
+    }
+
+    /**
+     * Expected for JDK lower to 7
+     */
+    private static boolean doesJDKDefineAnnotationsField() {
+        boolean jdkLower = true;
+        try {
+            Class.class.getDeclaredField(ANNOTATIONS_FIELD_NAME);
+        } catch (NoSuchFieldException ignore) {
+            jdkLower = false;
+        }
+        return jdkLower;
+    }
+
+    private static class LRAWrapped implements LRA {
```

Comment:
Fiendishly clever.

As an aside, we'd need to remember to change this class if the LRA annotation changes in MP-LRA 1.x.

Reply:
sure. I think that should be fine as this is part of the test which is usually compiled and run. The compilation would fail if the `LRAWrapped` does not implement some LRA method (or implement a different definition).